File: shared/UUID.erl
-module(shared_uUID@foreign).

-export([ eqUUIDImpl/2
        , compareUUIDImpl/5
        , stringToUUIDImpl/3
        , uuidToStringImpl/1
        , emptyImpl/0
        ]).

eqUUIDImpl(Lhs, Rhs) when is_binary(Lhs),
                          byte_size(Lhs) == 16 ->
  Lhs =:= Rhs.

compareUUIDImpl(Lt, Eq, Gt, Lhs, Rhs) when is_binary(Lhs),
                                           byte_size(Lhs) == 16,
                                           is_binary(Rhs),
                                           byte_size(Rhs) == 16
                                           ->
  if
    Lhs < Rhs -> Lt;
    Lhs =:= Rhs -> Eq;
    true -> Gt
  end.

stringToUUIDImpl(Nothing, Just, String) ->
  try
    Just(rtsv2_types:string_to_uuid(String))
  catch
    error:badarg ->
      Nothing
  end.

uuidToStringImpl(Uuid) ->
  rtsv2_types:uuid_to_string(Uuid).

emptyImpl() ->
  <<0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0>>.

File: shared/UUID.js
exports.eqUUIDImpl = function(lhs) {
  return function(rhs) {
    return lhs == rhs
  }
}
exports.compareUUIDImpl = function(lt) {
  return function(eq) {
    return function(gt) {
      return function(lhs) {
        return function(rhs) {
          if (lhs < rhs) {
            return lt;
          }
          else if (lhs == rhs) {
            return eq;
          }
          else {
            return gt;
          }
        }
      }
    }
  }
}
exports.stringToUUIDImpl = function(nothing) {
  return function(just) {
    return function(str) {
      if ((/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/).test(str)) {
        return just(str);
      }
      else {
        return nothing;
      }
    }
  }
}
exports.uuidToStringImpl = function(uuid) {
  return uuid;
}
exports.emptyImpl = function() {
  return "00000000-0000-0000-0000-000000000000";
}
